File: src/game/adventure/gameobjects/Room.java
package game.adventure.gameobjects;

import game.adventure.util.Constants;
import game.adventure.util.Direction;
import java.util.*;

public class Room extends GameObject {
    private final Map<Direction, String> directions;
    private final List<Item> items;
    private final List<Enemy> enemies;
    private boolean hasVisited;

    public Room(Builder builder) {
        super(builder);
        this.items = builder.items;
        this.enemies = builder.enemies;
        directions = new HashMap<>();
        for (Direction direction : Direction.values()) {
            directions.put(direction, "");
        }
    }

    public String getDirection(Direction direction) {
        return directions.get(direction);
    }

    public void setDirection(Direction direction, String north) {
        directions.put(direction, north);
    }

    public Map<Direction, String> getDirections() {
        return directions;
    }

    public Item getItem(int index) {
        return items.get(index);
    }

    public List<Item> getItems() {
        return items;
    }

    public void addItem(Item item) {
        items.add(item);
    }

    public void addItems(Item... items) {
        this.items.addAll(List.of(items));
    }

    public void removeItem(Item item) {
        items.remove(item);
    }

    public void clearItems() {
        items.clear();
    }

    public Item getItem(String itemName) {
        for (Item item : items) {
            if (item.getName().equalsIgnoreCase(itemName)) return item;
        }
        return Constants.GENERIC_ITEM;
    }

    public boolean hasItem(String item) {
        return !getItem(item).equals(Constants.GENERIC_ITEM);
    }

    public boolean hasItem(Item item) {
        return items.contains(item);
    }

    public boolean getHasVisited() {
        return hasVisited;
    }

    public void setHasVisited(boolean hasVisited) {
        this.hasVisited = hasVisited;
    }

    public List<Enemy> getEnemies() {
        return enemies;
    }

    public void clearEnemies() {
        enemies.clear();
    }

    public void addEnemy(Enemy enemy) {
        enemies.add(enemy);
    }

    public void removeEnemy(Enemy enemy) {
        enemies.remove(enemy);
    }

    public static class Builder extends GameObject.Builder<Builder> {
        private final List<Item> items = new ArrayList<>();
        private final List<Enemy> enemies = new ArrayList<>();

        public Builder items(List<Item> items) {
            this.items.addAll(items);
            return this;
        }

        public Builder items(Item... items) {
            this.items.addAll(List.of(items));
            return this;
        }

        public Builder enemies(List<Enemy> enemies) {
            this.enemies.addAll(enemies);
            return this;
        }

        public Builder enemies(Enemy... enemies) {
            this.enemies.addAll(List.of(enemies));
            return this;
        }

        @Override
        protected Builder self() {
            return this;
        }

        @Override
        public Room build() {
            return new Room(this);
        }
    }
}

File: src/game/adventure/gameobjects/Room.kt
package game.adventure.gameobjects

class Room(
    var name: String,
    var description: String,
    var n: Int,
    var s: Int,
    var e: Int,
    var w: Int,
    private val items: MutableList<Item>,
    private var enemies: MutableList<Enemy>
) {
    var hasVisited = false
    fun getItem(index: Int): Item {
        return items[index]
    }

    fun getItems(): List<Item> {
        return items
    }

    fun addItem(item: Item) {
        items.add(item)
    }

    fun removeItem(index: Int) {
        items.removeAt(index)
    }

    @Throws(IndexOutOfBoundsException::class)
    fun removeItem(item: String) {
        val index = getInt(item)
        if (index == -1) {
            throw IndexOutOfBoundsException()
        }
        items.remove(getItem(index))
    }

    fun clearItems() {
        items.clear()
    }

    fun removeItem(item: Item) {
        items.remove(item)
    }

    private fun getInt(name: String): Int {
        for (i in items.indices) {
            if (items[i].name.equals(name, ignoreCase = true)) {
                return i
            }
        }
        return -1
    }

    @Throws(IndexOutOfBoundsException::class)
    fun getItem(item: String): Item {
        val i = getInt(item)
        if (i == -1) {
            throw IndexOutOfBoundsException()
        }
        return getItems()[i]
    }

    fun hasItem(item: String): Boolean {
        return getInt(item) != -1
    }

    fun getEnemies(): MutableList<Enemy> {
        return enemies
    }

    fun setEnemies(enemies: MutableList<Enemy>) {
        this.enemies = enemies
    }

    fun addEnemy(enemy: Enemy) {
        enemies.add(enemy)
    }

    fun removeEnemy(enemy: Enemy) {
        enemies.remove(enemy)
    }
}